File: devtools/check_mastra_memory_replay.mjs
/** Provider-free agent entrypoint for check_mastra_memory_replay.py. */
import assert from "node:assert/strict";
import { readFile, writeFile } from "node:fs/promises";
import { createRequire } from "node:module";
import { join } from "node:path";
import { pathToFileURL } from "node:url";
import {
  createMemoryReplayAgent,
  createProcessLocalMemoryAccess,
} from "../packages/mastra/dist/memory.js";

const require = createRequire(
  new URL("../packages/mastra/package.json", import.meta.url),
);
const nativeImport = (name) =>
  import(pathToFileURL(require.resolve(name).replace(/\.cjs$/, ".js")).href);
const [{ InMemoryStore }, { Memory }, { z }] = await Promise.all([
  nativeImport("@mastra/core/storage"),
  nativeImport("@mastra/memory"),
  nativeImport("zod/v4"),
]);
const directory = process.env.CHECK_DIRECTORY;
const replayId = process.env.KITARU_REPLAY_ID;
const thread = "historical-thread";
const resource = "historical-resource";
const fileUrl = "https://files.invalid/historical.pdf?token=historical-secret";
const bytes = new Uint8Array(40000).fill(65);
const report = {
  task_id: process.env.KITARU_TASK_ID,
  replay_id: replayId ?? null,
  task_inputs_in_environment: process.env.KITARU_TASK_INPUTS !== undefined,
  source_calls: 0,
  file_calls: 0,
  dynamic_calls: 0,
  processor_calls: 0,
  actor_calls: 0,
};
assert(process.env.KITARU_API_TOKEN, "Expected the ordinary worker task token");
assert.equal(
  process.env.KITARU_API_KEY,
  undefined,
  "Administrative key leaked to task",
);
const denied = await fetch(
  `${process.env.KITARU_API_URL}/api/v1/blobs/${process.env.CHECK_UNRELATED_BLOB}/content`,
  {
    headers: { Authorization: `Bearer ${process.env.KITARU_API_TOKEN}` },
  },
);
assert.equal(denied.status, 403, "Task unexpectedly read an unrelated blob");
report.unrelated_blob_status = denied.status;
const stream = (parts, finishReason = "stop") => ({
  stream: new ReadableStream({
    start(controller) {
      controller.enqueue({ type: "stream-start", warnings: [] });
      for (const part of parts) controller.enqueue(part);
      controller.enqueue({
        type: "finish",
        finishReason,
        usage: { inputTokens: 5, outputTokens: 5, totalTokens: 10 },
      });
      controller.close();
    },
  }),
});
const text = (value) =>
  stream([
    { type: "text-start", id: "text" },
    { type: "text-delta", id: "text", delta: value },
    { type: "text-end", id: "text" },
  ]);
const actor = (id) => ({
  specificationVersion: "v2",
  supportedUrls: {},
  doGenerate: async () => {
    throw new Error("Unexpected generate call");
  },
  provider: "fixture",
  modelId: id,
  doStream: async (args) => {
    report.actor_calls++;
    const prompt = JSON.stringify(args.prompt);
    assert(prompt.includes("historical-blue"));
    assert(prompt.includes("Extra context"));
    assert(prompt.includes("HISTORICAL_SKILL"));
    assert(!prompt.includes("production-today"));
    if (replayId) {
      assert(prompt.includes("Replay application instruction"));
      assert(!prompt.includes("Original application instruction"));
      assert.equal(id, "replacement");
    }
    if (report.actor_calls === 1)
      return stream(
        [
          {
            type: "tool-call",
            toolCallId: "change-preference",
            toolName: "updateWorkingMemory",
            input: JSON.stringify({
              memory: {
                preference: replayId ? "replay-green" : "baseline-red",
              },
            }),
          },
        ],
        "tool-calls",
      );
    assert(prompt.includes(replayId ? "replay-green" : "baseline-red"));
    return text("done");
  },
});
const observationModel = (id) => ({
  specificationVersion: "v2",
  supportedUrls: {},
  provider: "fixture",
  modelId: id,
  doGenerate: async () => {
    throw new Error("Unexpected OM generate call");
  },
  doStream: async () => {
    throw new Error("Unexpected OM model call");
  },
});
const models = {
  actor: actor("actor"),
  replacement: actor("replacement"),
  observer: observationModel("observer"),
  reflector: observationModel("reflector"),
};
const store = new InMemoryStore();
const memory = new Memory({
  storage: store,
  options: {
    lastMessages: 10,
    semanticRecall: false,
    workingMemory: {
      enabled: true,
      scope: "thread",
      schema: z.object({ preference: z.string() }),
    },
    observationalMemory: {
      scope: "thread",
      observation: { model: "fixture/observer", messageTokens: 100000 },
      reflection: { model: "fixture/reflector", observationTokens: 100000 },
    },
  },
});
const domain = store.stores.memory;
assert(domain);
await memory.createThread({
  threadId: thread,
  resourceId: resource,
  title: "Historical thread",
});
const currentProduction = JSON.parse(
  await readFile(join(directory, "production.json"), "utf8"),
);
await memory.updateWorkingMemory({
  threadId: thread,
  resourceId: resource,
  workingMemory: JSON.stringify(currentProduction),
});
const nestedItems = Array.from({ length: 1500 }, (_, item) => ({
  id: item,
  details: Object.fromEntries(
    Array.from({ length: 10 }, (_, field) => [`field${field}`, field]),
  ),
}));
await domain.saveMessages({
  messages: Array.from({ length: 830 }, (_, index) => ({
    id: `historical-${index}`,
    role: "user",
    content: {
      format: 2,
      parts: [
        {
          type: "text",
          text: index === 0 ? "HISTORICAL_MESSAGE: " + "x".repeat(1_100_000) : `history ${index}`,
        },
      ],
      ...(index === 0 ? { metadata: { items: nestedItems } } : {}),
    },
    createdAt: new Date(Date.UTC(2026, 0, 1, 0, 0, index)),
    threadId: thread,
    resourceId: resource,
  })),
});
const omRecord = await domain.initializeObservationalMemory({
  threadId: thread,
  resourceId: resource,
  scope: "thread",
  config: { fixture: true },
});
await domain.updateActiveObservations({
  id: omRecord.id,
  observations: "HISTORICAL_OBSERVATION: likes blue.",
  tokenCount: 10,
  lastObservedAt: new Date(Date.UTC(2026, 0, 1, 0, 0, 0)),
  observedMessageIds: ["historical-0"],
});
const sourceDomain = new Proxy(domain, {
  get(target, key) {
    const value = Reflect.get(target, key, target);
    if (typeof value !== "function") return value;
    return (...args) => {
      if (replayId)
        throw new Error(
          `Production memory accessed during replay: ${String(key)}`,
        );
      return Reflect.apply(value, target, args);
    };
  },
});
const dynamic = (value) => () => {
  if (replayId)
    throw new Error("Live dynamic configuration was resolved during replay");
  report.dynamic_calls++;
  return value;
};
const agent = createMemoryReplayAgent(
  ({ memory: isolated, resolveFile, workspace }) => ({
    id: "headless-memory",
    name: "Headless memory",
    memory: isolated,
    workspace,
    instructions: dynamic("Original application instruction"),
    model: dynamic(models.actor),
    defaultOptions: dynamic({ maxSteps: 3 }),
    inputProcessors: [
      {
        id: "historical-file",
        async processInput({ messages }) {
          report.processor_calls++;
          const filePart = messages
            .flatMap((message) => message.content.parts)
            .find((part) => part.type === "file");
          assert(filePart, "Expected a file in native processor input");
          const file = await resolveFile(String(filePart.data));
          assert.deepEqual(file.bytes, bytes);
          return messages.map((message) => ({
            ...message,
            content: {
              ...message.content,
              parts: message.content.parts.map((part) =>
                part.type === "file"
                  ? {
                      ...part,
                      data: Buffer.from(file.bytes).toString("base64"),
                    }
                  : part,
              ),
            },
          }));
        },
      },
    ],
  }),
  {
    agentId: process.env.CHECK_AGENT_ID,
    requestedModelId: "fixture/actor",
    allowedReplayModels: ["fixture/replacement"],
    sourceMemory: () => {
      report.source_calls++;
      if (replayId)
        throw new Error("Production source requested during replay");
      return {
        domain: sourceDomain,
        configuration: memory.getMergedThreadConfig(),
        settled: () => memory.settled(),
        exclusiveAccess: createProcessLocalMemoryAccess(),
      };
    },
    resolveModel: (id) => {
      const model = models[id.split("/").at(-1)];
      assert(model, `Unexpected model ${id}`);
      return model;
    },
    files: [fileUrl],
    resolveFile: async () => {
      report.file_calls++;
      if (replayId) throw new Error("Live file requested during replay");
      return { bytes, mediaType: "application/pdf" };
    },
    skillsDirectory: join(directory, "skills"),
  },
);
try {
  const result = await agent.stream(
    [
      {
        role: "user",
        content: [
          { type: "text", text: "Update the preference using the attachment." },
          { type: "file", data: new URL(fileUrl), mimeType: "application/pdf" },
        ],
      },
    ],
    {
      memory: { thread, resource },
      context: [{ role: "system", content: "Extra context" }],
    },
  );
  await result.consumeStream();
  assert.equal(await result.text, "done");
  assert.equal(report.processor_calls, 1);
  assert.equal(report.actor_calls, 2);
  assert.equal(report.dynamic_calls, replayId ? 0 : 3);
  assert.equal(report.source_calls, replayId ? 0 : 1);
  assert.equal(report.file_calls, replayId ? 0 : 1);
  if (replayId) {
    assert.equal(report.task_inputs_in_environment, false);
    assert.deepEqual(
      JSON.parse(await readFile(join(directory, "production.json"), "utf8")),
      { preference: "production-today" },
    );
    assert(
      (
        await memory.getWorkingMemory({
          threadId: thread,
          resourceId: resource,
        })
      ).includes("production-today"),
    );
  }
  report.result = "passed";
} catch (error) {
  report.result = "failed";
  report.error = String(error);
  throw error;
} finally {
  await memory.settled();
  await store.close();
  await writeFile(
    join(directory, `${report.task_id}.json`),
    JSON.stringify(report, null, 2),
  );
}

File: devtools/check_mastra_memory_replay.py
"""Prove SDK, CLI and MCP memory replay on an authenticated disposable stack.

Run ``pnpm run build:packages`` first, then
``uv run python devtools/check_mastra_memory_replay.py``.
No model provider is called. Only this invocation's server and database are removed.
"""

import argparse
import asyncio
import copy
import json
import os
import shutil
import subprocess
import sys
import tempfile
import uuid
from pathlib import Path
from typing import Any

import asyncpg
from mcp.client.session import ClientSession
from mcp.client.stdio import StdioServerParameters, stdio_client
from stack import (
    DB_HOST,
    DB_PORT,
    DB_PWD,
    DB_USER,
    bootstrap_api_key,
    create_database,
    drop_database,
    ensure_postgres,
    get_free_port,
    start_server,
    wait_for_health,
)

from kitaru.api_models.v1.agent import AgentCreateRequest
from kitaru.api_models.v1.agent_version import AgentVersionCreateRequest, RunSpec
from kitaru.api_models.v1.evaluator import (
    EvaluatorCreateRequest,
    EvaluatorVersionCreateRequest,
)
from kitaru.api_models.v1.job import JobResponse, JobStatus
from kitaru.api_models.v1.plugin import EvaluatorConfig, ScriptPluginSource
from kitaru.api_models.v1.replay import ReplayCreateRequest
from kitaru.api_models.v1.replay_config import (
    ReplayOverride,
    ToolPolicy,
)
from kitaru.api_models.v1.session import (
    SessionCreateRequest,
    SessionOrigin,
    SessionStatus,
)
from kitaru.api_models.v1.session_node import SessionNodeListParams
from kitaru.api_models.v1.session_run import SessionRunCreateRequest
from kitaru.client.api_client import KitaruAPIClient
from kitaru.client.exceptions import APIError
from kitaru.worker import Worker, WorkerConfig

ARTIFACT = Path(__file__).with_suffix(".mjs").resolve()
OVERRIDE = {
    "system_prompt": "Replay application instruction",
    "model": "fixture/replacement",
}
POLICY = {
    "default": {"type": "history", "scope": "baseline", "on_miss": "fail"},
    "tools": {},
}


async def await_job(
    client: KitaruAPIClient, job_id: uuid.UUID, label: str, timeout: float
) -> JobResponse:
    """Wait for one owned job to reach a terminal state."""
    deadline = asyncio.get_running_loop().time() + timeout
    while True:
        job = await client.jobs.get(job_id)
        if job.status in {JobStatus.COMPLETED, JobStatus.FAILED, JobStatus.CANCELED}:
            return job
        if asyncio.get_running_loop().time() >= deadline:
            raise TimeoutError(f"{label} did not complete within {timeout}s")
        await asyncio.sleep(0.2)


async def call_mcp(session: ClientSession, tool: str, request: dict[str, Any]) -> Any:
    """Call a public MCP tool and require its structured success envelope."""
    result = await session.call_tool(tool, {"request": request})
    assert not result.is_error, result
    envelope = result.structured_content
    assert envelope and envelope["ok"], envelope
    return envelope["data"]


async def run_cli(*args: str) -> dict[str, Any]:
    """Run the installed CLI without blocking the worker's event loop."""
    process = await asyncio.create_subprocess_exec(
        str(Path(sys.executable).with_name("kitaru")),
        *args,
        "--output",
        "json",
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    try:
        stdout, stderr = await asyncio.wait_for(process.communicate(), 60)
    except BaseException:
        if process.returncode is None:
            process.kill()
            await process.wait()
        raise
    assert process.returncode == 0, stderr.decode() + stdout.decode()
    return json.loads(stdout)


async def inspect_session(
    client: KitaruAPIClient, session_id: uuid.UUID
) -> dict[str, Any]:
    """Require complete, hydrated request evidence and ordered memory changes."""
    session = await client.sessions.get(session_id)
    assert session.status == SessionStatus.COMPLETED, session.error
    nodes = [
        node
        async for node in client.sessions.iter_nodes(
            session_id, SessionNodeListParams(include_payloads=True, size=1)
        )
    ]
    model_nodes = [node for node in nodes if node.node_type == "llm_call"]
    mutations = [node for node in nodes if node.name == "memory_mutation"]
    assert len(model_nodes) == 2, [(node.name, node.node_type) for node in nodes]
    assert mutations, "Memory changes were not recorded"
    revisions = [node.attributes["memory_revision"] for node in mutations]
    assert revisions == sorted(set(revisions)), revisions
    assert all(node.attributes["evidence_complete"] for node in mutations)
    request_ids = {node.external_id for node in model_nodes}
    assert any(node.attributes["request_id"] in request_ids for node in mutations)
    assert all(node.attributes["request_complete"] for node in model_nodes)
    assert all(len(json.dumps(node.inputs)) > 40000 for node in model_nodes)
    first_request = json.dumps(model_nodes[0].inputs)
    assert "historical-blue" in first_request
    assert "production-today" not in first_request
    assert "Replay application instruction" in first_request
    assert "Extra context" in first_request
    assert "HISTORICAL_SKILL" in first_request
    assert "replay-green" in json.dumps(model_nodes[1].inputs)
    assert all(node.model == "replacement" for node in model_nodes)
    return {
        "session_id": str(session_id),
        "task_id": str(session.task_id),
        "node_ids": [str(node.id) for node in nodes],
        "model_nodes": len(model_nodes),
        "memory_nodes": len(mutations),
    }


async def finish_replay(
    client: KitaruAPIClient, replay_id: uuid.UUID
) -> dict[str, Any]:
    """Wait for both replay and evaluator completion, then inspect the result."""
    replay = await client.replays.get(replay_id)
    assert replay.job_id
    job = await await_job(client, replay.job_id, "memory replay", 120)
    if job.status != JobStatus.COMPLETED:
        tasks = await client.jobs.list_tasks(job.id)
        raise AssertionError(
            [(task.kind, task.status, task.error) for task in tasks.items]
        )
    replay = await client.replays.get(replay_id)
    assert replay.status == "completed", replay
    assert replay.result_session_id
    return {
        "replay_id": str(replay.id),
        **await inspect_session(client, replay.result_session_id),
    }


async def check_mcp(
    client: KitaruAPIClient,
    url: str,
    directory: Path,
    agent_id: uuid.UUID,
    version_id: uuid.UUID,
    evaluator_id: uuid.UUID,
    baseline_id: uuid.UUID,
) -> dict[str, Any]:
    """Create an experiment and inspect every result page over real MCP stdio."""
    mcp_console = str(Path(sys.executable).with_name("kitaru-mcp"))
    with (directory / "mcp.log").open("w") as mcp_log:
        parameters = StdioServerParameters(
            command=mcp_console,
            args=["--server", url, "--mode", "standard"],
            env=dict(os.environ),
        )
        async with (
            stdio_client(parameters, errlog=mcp_log) as (reader, writer),
            ClientSession(reader, writer) as mcp,
        ):
            await mcp.initialize()
            cohort = await call_mcp(
                mcp,
                "kitaru_cohorts_manage",
                {
                    "operation": "create",
                    "agent_id": str(agent_id),
                    "name": "memory-cohort",
                },
            )
            cohort_version = await call_mcp(
                mcp,
                "kitaru_cohorts_manage",
                {
                    "operation": "create_version",
                    "cohort_id": cohort["id"],
                    "add_session_ids": [str(baseline_id)],
                },
            )
            experiment = await call_mcp(
                mcp,
                "kitaru_experiments_manage",
                {
                    "operation": "create",
                    "agent_id": str(agent_id),
                    "name": "memory-experiment",
                    "override": OVERRIDE,
                    "tool_policy": POLICY,
                    "evaluators": [{"evaluator_id": str(evaluator_id), "version": 1}],
                },
            )
            started = await call_mcp(
                mcp,
                "kitaru_workflow_start",
                {
                    "operation": "experiment_run",
                    "experiment_id": experiment["id"],
                    "cohort_version_id": cohort_version["id"],
                    "agent_version_id": str(version_id),
                    "baseline_evaluation_mode": "none",
                },
            )
            run_id = started["result"]["id"]
            deadline = asyncio.get_running_loop().time() + 120
            while True:
                run = await call_mcp(
                    mcp,
                    "kitaru_activity_read",
                    {
                        "operation": "get",
                        "kind": "experiment_run",
                        "id": run_id,
                    },
                )
                if run["status"] in {"completed", "failed", "canceled"}:
                    break
                assert asyncio.get_running_loop().time() < deadline, (
                    "MCP experiment timed out"
                )
                await asyncio.sleep(0.2)
            assert run["status"] == "completed", run
            children = await call_mcp(
                mcp,
                "kitaru_activity_read",
                {
                    "operation": "list_children",
                    "kind": "experiment_run_jobs",
                    "parent_id": run_id,
                    "size": 1,
                },
            )
            assert len(children["items"]) == 1
            replay_page = await call_mcp(
                mcp,
                "kitaru_activity_read",
                {
                    "operation": "list",
                    "kind": "replay",
                    "size": 1,
                    "filter": {
                        "field": "experiment_run_id",
                        "op": "eq",
                        "value": run_id,
                    },
                },
            )
            assert len(replay_page["items"]) == 1
            mcp_replay = replay_page["items"][0]
            assert mcp_replay["status"] == "completed"
            mcp_result = await finish_replay(client, uuid.UUID(mcp_replay["id"]))
            mcp_session = await call_mcp(
                mcp,
                "kitaru_activity_read",
                {
                    "operation": "get",
                    "kind": "session",
                    "id": mcp_result["session_id"],
                },
            )
            assert mcp_session["status"] == "completed"
            assert mcp_session["inputs"]["mastra_memory_replay"]["complete"]
            mcp_nodes = []
            cursor = None
            while True:
                page = await call_mcp(
                    mcp,
                    "kitaru_activity_read",
                    {
                        "operation": "list_children",
                        "kind": "session_nodes",
                        "parent_id": mcp_result["session_id"],
                        "size": 1,
                        "cursor": cursor,
                        "include_payloads": True,
                    },
                )
                mcp_nodes.extend(page["items"])
                if not page["page"]["has_more"]:
                    break
                cursor = page["page"]["next_cursor"]
                assert cursor
            assert {item["id"] for item in mcp_nodes} == set(mcp_result["node_ids"])
            assert all(
                item["inputs"] and item["attributes"]["request_complete"]
                for item in mcp_nodes
                if item["node_type"] == "llm_call"
            )
            assert any(
                item["name"] == "memory_mutation"
                and item["attributes"]["memory_revision"] > 0
                for item in mcp_nodes
            )
        parameters = StdioServerParameters(
            command=mcp_console,
            args=["--server", url, "--mode", "read-only"],
            env=dict(os.environ),
        )
        async with (
            stdio_client(parameters, errlog=mcp_log) as (reader, writer),
            ClientSession(reader, writer) as mcp,
        ):
            await mcp.initialize()
            names = {tool.name for tool in (await mcp.list_tools()).tools}
            assert "kitaru_workflow_start" not in names
            assert "kitaru_experiments_manage" not in names
            read = await call_mcp(
                mcp,
                "kitaru_activity_read",
                {
                    "operation": "get",
                    "kind": "replay",
                    "id": mcp_result["replay_id"],
                },
            )
            assert read["result_session_id"] == mcp_result["session_id"]
            refused = await mcp.call_tool(
                "kitaru_workflow_start",
                {
                    "request": {
                        "operation": "experiment_run",
                        "experiment_id": experiment["id"],
                        "cohort_version_id": cohort_version["id"],
                        "agent_version_id": str(version_id),
                    }
                },
            )
            assert refused.is_error
    print("MCP experiment, pagination and read-only mode passed", flush=True)
    return {
        "experiment_run_id": run_id,
        **mcp_result,
        "paginated_nodes": len(mcp_nodes),
    }


async def check(output: Path) -> None:
    """Exercise all headless paths and clean up owned resources even on failure."""
    node = shutil.which("node")
    assert node, "Node 22.22 or 26 is required"
    db_name = f"kitaru_memory_{uuid.uuid4().hex[:10]}"
    directory = Path(tempfile.mkdtemp(prefix="kitaru-memory-proof-"))
    print(f"Proof logs: {directory}", flush=True)
    skills = directory / "skills" / "triage"
    skills.mkdir(parents=True)
    (skills / "SKILL.md").write_text(
        "---\nname: triage\ndescription: HISTORICAL_SKILL.\n---\n"
        "Use historical knowledge.\n"
    )
    (directory / "production.json").write_text(
        json.dumps({"preference": "historical-blue"})
    )
    stop = asyncio.Event()
    server = None
    worker_task = None
    await ensure_postgres()
    await create_database(db_name)
    try:
        port = get_free_port()
        url = f"http://127.0.0.1:{port}"
        server = start_server(
            db_name, port, directory / "server.log", auth_scheme="local"
        )
        await wait_for_health(url, server, directory / "server.log")
        key = await bootstrap_api_key(url)
        assert key
        os.environ["KITARU_API_URL"] = url
        os.environ["KITARU_API_KEY"] = key
        os.environ.pop("KITARU_API_TOKEN", None)
        worker = Worker(
            WorkerConfig(
                name=f"memory-proof-{uuid.uuid4().hex[:8]}",
                concurrency=1,
                poll_interval=0.1,
                heartbeat_interval=0.5,
                blob_cache_root=directory / "worker-blobs",
                payload_cache_root=directory / "worker-payloads",
            )
        )
        worker_task = asyncio.create_task(worker.run(stop))
        async with KitaruAPIClient(base_url=url, api_key=key) as client:
            unrelated = await client.blobs.upload(
                b"unrelated private content", media_type="text/plain"
            )
            agent = await client.agents.create(AgentCreateRequest(name="memory-proof"))
            version = await client.agents.create_version(
                agent.id,
                AgentVersionCreateRequest(
                    run_spec=RunSpec(
                        command=f'"{node}" "{ARTIFACT}"',
                        timeout_seconds=90,
                        env={
                            "CHECK_AGENT_ID": str(agent.id),
                            "CHECK_DIRECTORY": str(directory),
                            "CHECK_UNRELATED_BLOB": str(unrelated.id),
                        },
                    )
                ),
            )
            job = await client.session_runs.create(
                SessionRunCreateRequest(
                    agent_version_id=version.id,
                    inputs=[
                        {
                            "role": "user",
                            "content": [
                                {
                                    "type": "text",
                                    "text": "Update the preference.",
                                },
                                {
                                    "type": "file",
                                    "data": "https://files.invalid/historical.pdf?token=historical-secret",
                                    "mimeType": "application/pdf",
                                },
                            ],
                        }
                    ],
                )
            )
            job = await await_job(client, job.id, "baseline", 120)
            if job.status != JobStatus.COMPLETED:
                tasks = await client.jobs.list_tasks(job.id)
                raise AssertionError(
                    [(task.kind, task.status, task.error) for task in tasks.items]
                )
            sessions = [session async for session in client.sessions.iter()]
            assert len(sessions) == 1
            deadline = asyncio.get_running_loop().time() + 30
            while True:
                baseline = await client.sessions.get(sessions[0].id)
                if baseline.metadata.get("mastra_replay_state") != "pending":
                    break
                assert asyncio.get_running_loop().time() < deadline, (
                    "Baseline memory evidence did not finalize"
                )
                await asyncio.sleep(0.2)
            assert baseline.metadata["mastra_replay_state"] == "eligible"
            envelope = baseline.inputs["mastra_memory_replay"]
            assert envelope["complete"] is True
            assert len(envelope["initialSnapshot"]["messages"]) == 830
            assert envelope["initialSnapshot"]["records"]
            first_message = next(
                message
                for message in envelope["initialSnapshot"]["messages"]
                if message["id"] == "historical-0"
            )
            hotels = first_message["content"]["metadata"]["hotels"]
            assert len(hotels) == 1500
            assert all(len(hotel["details"]) == 10 for hotel in hotels)
            assert len(json.dumps(baseline.inputs).encode()) > 1_048_576
            assert "historical-secret" not in json.dumps(baseline.inputs)
            # Recorded files live in blobs, outside the replay input.
            assert envelope["files"]
            assert all("base64" not in file for file in envelope["files"])
            file_blob_ids = [file["blobId"] for file in envelope["files"]]
            connection = await asyncpg.connect(
                host=DB_HOST,
                port=DB_PORT,
                user=DB_USER,
                password=DB_PWD,
                database=db_name,
            )
            try:
                offloaded = await connection.fetchval(
                    "SELECT inputs_blob_id FROM session WHERE id = $1", baseline.id
                )
                assert offloaded, (
                    "Large session inputs did not use server payload offload"
                )
            finally:
                await connection.close()
            (directory / "production.json").write_text(
                json.dumps({"preference": "production-today"})
            )
            evaluator_blob = await client.blobs.upload(
                (Path(__file__).parent / "evaluators.py").read_bytes(),
                media_type="text/x-python",
                filename="evaluator.py",
            )
            evaluator = await client.evaluators.create(
                EvaluatorCreateRequest(name="memory-completed")
            )
            await client.evaluators.create_version(
                evaluator.id,
                EvaluatorVersionCreateRequest(
                    source=ScriptPluginSource(
                        blob_id=evaluator_blob.id, entrypoint="evaluate_outcome"
                    )
                ),
            )
            config = EvaluatorConfig(evaluator="memory-completed", version=1)
            replay = await client.replays.create(
                ReplayCreateRequest(
                    baseline_session_id=baseline.id,
                    override=ReplayOverride(**OVERRIDE),
                    tool_policy=ToolPolicy.model_validate(POLICY),
                    evaluators=[config],
                )
            )
            sdk = await finish_replay(client, replay.id)
            print("SDK replay passed", flush=True)
            created = await run_cli(
                "replay",
                "create",
                str(baseline.id),
                "--evaluator",
                "memory-completed@1",
                "--override",
                json.dumps(OVERRIDE),
                "--tool-policy",
                json.dumps(POLICY),
            )
            cli = await finish_replay(client, uuid.UUID(created["item"]["id"]))
            cli_read = await run_cli("replay", "get", cli["replay_id"])
            assert cli_read["item"]["status"] == "completed"
            assert cli_read["item"]["result_session_id"] == cli["session_id"]
            cli_nodes = await run_cli(
                "session",
                "nodes",
                cli["session_id"],
                "--include-payloads",
                "--size",
                "100",
            )
            assert {item["id"] for item in cli_nodes["items"]} == set(cli["node_ids"])
            assert all(
                item["inputs"] and item["attributes"]["request_complete"]
                for item in cli_nodes["items"]
                if item["node_type"] == "llm_call"
            )
            assert any(item["name"] == "memory_mutation" for item in cli_nodes["items"])
            print("CLI replay passed", flush=True)
            mcp_result = await check_mcp(
                client, url, directory, agent.id, version.id, evaluator.id, baseline.id
            )
            incomplete = copy.deepcopy(baseline.inputs)
            incomplete["mastra_memory_replay"]["complete"] = False
            broken = await client.sessions.create(
                SessionCreateRequest(
                    agent_id=agent.id,
                    agent_version_id=version.id,
                    framework="mastra",
                    origin=SessionOrigin.RECORDED,
                    status=SessionStatus.COMPLETED,
                    inputs=incomplete,
                    outputs={"text": "incomplete fixture"},
                )
            )
            try:
                await client.replays.create(
                    ReplayCreateRequest(
                        baseline_session_id=broken.id,
                        evaluators=[config],
                        override=ReplayOverride(**OVERRIDE),
                    )
                )
            except APIError as error:
                assert error.status_code == 409
                assert "mastra_replay_incomplete" in error.detail
            else:
                raise AssertionError("Incomplete replay was scheduled")
            reports = [
                json.loads(path.read_text())
                for path in directory.glob("*.json")
                if path.name != "production.json"
            ]
            good = [report for report in reports if report.get("result") == "passed"]
            assert len(good) == 4, reports
            assert not [
                report for report in reports if report.get("result") == "failed"
            ]
            assert all(report["unrelated_blob_status"] == 403 for report in good)
            assert all(
                not report["task_inputs_in_environment"]
                for report in good
                if report["replay_id"]
            )
            server_log = (directory / "server.log").read_text()
            for report in good:
                expected = 1 if report["replay_id"] else 0
                report["task_spec_requests"] = server_log.count(
                    f'"GET /api/v1/tasks/{report["task_id"]}/spec HTTP/1.1" 200 OK'
                )
                assert report["task_spec_requests"] == expected, report
            # Each replay task read the recorded files with its own task token.
            replays = sum(1 for report in good if report["replay_id"])
            file_blob_downloads = {
                blob_id: server_log.count(
                    f'"GET /api/v1/blobs/{blob_id}/content HTTP/1.1" 200 OK'
                )
                for blob_id in file_blob_ids
            }
            assert all(count == replays for count in file_blob_downloads.values()), (
                file_blob_downloads
            )
            connection = await asyncpg.connect(
                host=DB_HOST,
                port=DB_PORT,
                user=DB_USER,
                password=DB_PWD,
                database=db_name,
            )
            try:
                offloaded_requests = await connection.fetchval(
                    "SELECT count(*) FROM session_node "
                    "WHERE session_id = ANY($1::uuid[]) "
                    "AND node_type = 'llm_call' AND inputs_blob_id IS NOT NULL",
                    [uuid.UUID(item["session_id"]) for item in (sdk, cli, mcp_result)],
                )
                assert offloaded_requests == 6, offloaded_requests
            finally:
                await connection.close()
            proof = {
                "result": "passed",
                "baseline_session_id": str(baseline.id),
                "offloaded_input_blob_id": str(offloaded),
                "offloaded_request_nodes": offloaded_requests,
                "recorded_file_blob_downloads": file_blob_downloads,
                "sdk": sdk,
                "cli": cli,
                "mcp": mcp_result,
                "incomplete_baseline_id": str(broken.id),
                "read_only_mutation_denied": True,
                "task_reports": reports,
                "provider_calls": 0,
                "logs": str(directory),
            }
            output.parent.mkdir(parents=True, exist_ok=True)
            output.write_text(json.dumps(proof, indent=2) + "\n")
            print(
                f"PASS: authenticated memory replay proof saved to {output}", flush=True
            )
    finally:
        stop.set()
        if worker_task is not None:
            try:
                await asyncio.wait_for(
                    asyncio.gather(worker_task, return_exceptions=True), 15
                )
            except TimeoutError:
                worker_task.cancel()
                await asyncio.gather(worker_task, return_exceptions=True)
        if server is not None:
            server.terminate()
            try:
                server.wait(timeout=10)
            except subprocess.TimeoutExpired:
                server.kill()
                server.wait(timeout=10)
        await drop_database(db_name)
        for name in ("worker-blobs", "worker-payloads"):
            shutil.rmtree(directory / name, ignore_errors=True)
        print(
            "Removed the owned server, worker, database and worker caches.", flush=True
        )


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--output", type=Path, default=Path("/tmp/kitaru-memory-proof.json")
    )
    asyncio.run(check(parser.parse_args().output))
